File: tests/test_rule_schema.js
import assert from 'node:assert/strict';
import test from 'node:test';
import { JS_HANDLERS, loadMoneylineRules } from '../src/rule_engine.js';

const VALID_ACTIONS = new Set(['NO_BET', 'CAP', 'ADJUST']);
const VALID_ENGINES = new Set(['js', 'py']);
const KNOWN_HANDLERS = new Set(); // handlers referenced by py-only rules; not registered in JS

test('rules file has the pinned version', () => {
  const rules = loadMoneylineRules();
  assert.equal(rules.version, 'moneyline-rules-v1');
});

test('every rule has required, well-typed fields', () => {
  const { rules } = loadMoneylineRules();
  assert.ok(Array.isArray(rules) && rules.length > 0);
  for (const rule of rules) {
    assert.equal(typeof rule.id, 'string', `id missing on ${JSON.stringify(rule)}`);
    assert.ok(Array.isArray(rule.engines) && rule.engines.length > 0, `engines missing on ${rule.id}`);
    assert.ok(rule.engines.every((e) => VALID_ENGINES.has(e)), `bad engine on ${rule.id}`);
    assert.ok([1, 2, 3].includes(rule.tier), `tier must be 1/2/3 on ${rule.id}`);
    assert.equal(typeof rule.order, 'number', `order missing on ${rule.id}`);
    assert.ok(VALID_ACTIONS.has(rule.action), `bad action on ${rule.id}`);
    assert.equal(typeof rule.handler, 'string', `handler missing on ${rule.id}`);
    assert.equal(typeof rule.message, 'string', `message missing on ${rule.id}`);
  }
});

test('rule ids are globally unique', () => {
  const { rules } = loadMoneylineRules();
  const ids = rules.map((r) => r.id);
  assert.equal(new Set(ids).size, ids.length);
});

test('(engine, order) pairs are unique within each engine', () => {
  const { rules } = loadMoneylineRules();
  for (const engine of VALID_ENGINES) {
    const orders = rules.filter((r) => r.engines.includes(engine)).map((r) => r.order);
    assert.equal(new Set(orders).size, orders.length, `duplicate order in engine ${engine}`);
  }
});

test('JS-scoped rule handlers all exist in JS_HANDLERS (no orphan rules)', () => {
  const { rules } = loadMoneylineRules();
  for (const rule of rules) {
    if (!rule.engines.includes('js')) continue;
    assert.ok(
      typeof JS_HANDLERS[rule.handler] === 'function',
      `JS rule ${rule.id} references unregistered handler ${rule.handler}`
    );
  }
});

test('every registered JS_HANDLER is referenced by a JS-scoped rule (no dead handlers)', () => {
  const { rules } = loadMoneylineRules();
  const referenced = new Set(
    rules.filter((r) => r.engines.includes('js')).map((r) => r.handler)
  );
  for (const name of Object.keys(JS_HANDLERS)) {
    if (KNOWN_HANDLERS.has(name)) continue;
    assert.ok(referenced.has(name), `JS_HANDLERS.${name} is not referenced by any JS rule`);
  }
});

File: tests/test_rule_schema.py
"""Schema/invariant tests for data/rules/moneyline_rules.json (Python view).

Mirrors tests/test_rule_schema.js. Enforces the pinned version, well-typed
fields, globally-unique ids, unique (engine, order) pairs, and a strict
bijection between py-scoped rule handlers and PY_HANDLERS (no orphan rules, no
dead handlers).
"""

import unittest

from src.rule_engine import PY_HANDLERS, load_moneyline_rules

VALID_ACTIONS = {"NO_BET", "CAP", "ADJUST"}
VALID_ENGINES = {"js", "py"}


class RuleSchemaTest(unittest.TestCase):
    def setUp(self) -> None:
        self.catalog = load_moneyline_rules()
        self.rules = self.catalog["rules"]

    def test_pinned_version(self) -> None:
        self.assertEqual(self.catalog["version"], "moneyline-rules-v1")

    def test_every_rule_has_well_typed_fields(self) -> None:
        self.assertTrue(isinstance(self.rules, list) and len(self.rules) > 0)
        for rule in self.rules:
            rid = rule.get("id")
            self.assertIsInstance(rid, str, f"id missing on {rule}")
            self.assertTrue(
                isinstance(rule.get("engines"), list) and rule["engines"],
                f"engines missing on {rid}",
            )
            self.assertTrue(
                all(e in VALID_ENGINES for e in rule["engines"]),
                f"bad engine on {rid}",
            )
            self.assertIn(rule.get("tier"), (1, 2, 3), f"tier must be 1/2/3 on {rid}")
            self.assertIsInstance(rule.get("order"), (int, float), f"order missing on {rid}")
            self.assertIn(rule.get("action"), VALID_ACTIONS, f"bad action on {rid}")
            self.assertIsInstance(rule.get("handler"), str, f"handler missing on {rid}")
            self.assertIsInstance(rule.get("message"), str, f"message missing on {rid}")

    def test_rule_ids_are_globally_unique(self) -> None:
        ids = [rule["id"] for rule in self.rules]
        self.assertEqual(len(set(ids)), len(ids))

    def test_engine_order_pairs_unique_within_each_engine(self) -> None:
        for engine in VALID_ENGINES:
            orders = [rule["order"] for rule in self.rules if engine in rule["engines"]]
            self.assertEqual(
                len(set(orders)), len(orders), f"duplicate order in engine {engine}"
            )

    def test_py_rule_handlers_all_exist(self) -> None:
        for rule in self.rules:
            if "py" not in rule["engines"]:
                continue
            self.assertIn(
                rule["handler"],
                PY_HANDLERS,
                f"Python rule {rule['id']} references unregistered handler {rule['handler']}",
            )

    def test_every_registered_handler_is_referenced(self) -> None:
        referenced = {
            rule["handler"] for rule in self.rules if "py" in rule["engines"]
        }
        for name in PY_HANDLERS:
            self.assertIn(
                name, referenced, f"PY_HANDLERS.{name} is not referenced by any py rule"
            )


if __name__ == "__main__":
    unittest.main()
